File: 0001-0099/0017. Letter Combinations of a Phone Number/0017. Letter Combinations of a Phone Number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
  const ans = [];
  if (!digits.length) return ans;
  ans.push("");
  const mapping = ["0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
  while (ans[0].length !== digits.length) {
    const t = ans.shift();
    for (const item of [...mapping[digits[t.length]]]) {
      ans.push(t + item);
    }
  }
  return ans;
};
letterCombinations("23");

File: 0001-0099/0017. Letter Combinations of a Phone Number/0017. Letter Combinations of a Phone Number.py
from typing import List


class Solution:

    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        mapping = ['0', '1', "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        queue = ['']
        for item in digits:
            for _ in range(len(queue)):
                tmp = queue.pop(0)
                for letter in mapping[int(item)]:
                    queue.append(tmp + letter)
        return queue
